File: problems/046/046.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */
/* File Reference: http://www.naturalnumbers.org/primes.html */

int main()
{
    int smallest_odd_composite_number = 0;

    const int MAX_N = 1E5;

    FILE *fp;
    char prime_num[10];
    const char* FILE_NAME = "problems/003/PrimeNumbers_Upto_1000000";
    fp = fopen(FILE_NAME, "r");

    int primes_list[MAX_N+1];
    memset(&primes_list[0], 0, sizeof(primes_list));

    while (fgets(prime_num, 10, fp) != NULL)
    {
        if (atoi(prime_num) > MAX_N)
            break;

        primes_list[atoi(prime_num)] = 1;
    }
    fclose(fp);

    for (int i = 35; i < MAX_N; i += 2)
    {
        if (primes_list[i])
            continue;

        int j = 1;
        int prime_found = 0;
        while (++j < i)
        {
            if (!primes_list[j])
                continue;

            float b = sqrt((float)(i - j) / 2);

            if (b == (int)b)
            {
                prime_found = 1;
                break;
            }
        }

        if (!prime_found)
        {
            smallest_odd_composite_number = i;
            break;
        }
    }

    printf("%d\n", smallest_odd_composite_number);

    return 0;
}

File: problems/046/046.f90
PROGRAM F046

    ! Copyright 2021 Melwyn Francis Carlo
    ! FILE REFERENCE: http://www.naturalnumbers.org/primes.html

    IMPLICIT NONE

    INTEGER, PARAMETER :: MAX_N = 1E5

    INTEGER :: PRIME_NUM
    CHARACTER (LEN=50), PARAMETER :: FILENAME = "problems/003/PrimeNumbers_Upto_1000000"

    INTEGER, DIMENSION(MAX_N)     :: PRIMES_LIST

    INTEGER :: SMALLEST_ODD_COMPOSITE_NUMBER = 0
    INTEGER :: I, J

    REAL    :: B

    LOGICAL :: PRIME_FOUND

    OPEN (2, FILE = FILENAME, STATUS = 'OLD')

    DO
        READ(2, *, END=10) PRIME_NUM
        IF (PRIME_NUM > MAX_N) EXIT
        PRIMES_LIST(PRIME_NUM) = 1
    END DO

    10 CLOSE(2)

    DO I = 35, MAX_N, 2

        IF (PRIMES_LIST(I) /= 0) CYCLE

        J = 2
        PRIME_FOUND = .FALSE.
        DO WHILE (J < I)

            IF (PRIMES_LIST(J) == 0) THEN
                J = J + 1
                CYCLE
            END IF

            B = SQRT(FLOAT(I - J) / 2.0)

            IF (B - REAL(INT(B)) <= 0) THEN
                PRIME_FOUND = .TRUE.
                EXIT
            END IF

            J = J + 1

        END DO

        IF (.NOT. PRIME_FOUND) THEN
            SMALLEST_ODD_COMPOSITE_NUMBER = I
            EXIT
        END IF

    END DO

    PRINT ('(I0)'), SMALLEST_ODD_COMPOSITE_NUMBER

END PROGRAM F046
